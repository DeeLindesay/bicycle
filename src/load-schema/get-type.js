import assert from 'assert';
import freeze from 'bicycle/utils/freeze';

function getType(strType: string, context?: string): Object {
  if (strType[strType.length - 1] !== '?') {
    return freeze({kind: 'NotNull', type: getType(strType + '?')});
  }
  strType = strType.substr(0, strType.length - 1);
  if (strType[strType.length - 2] === '[' && strType[strType.length - 1] === ']') {
    return freeze({kind: 'List', type: getType(strType.substr(0, strType.length - 2))});
  }
  const contextStr = context ? ' for ' + context : '';
  assert(
    /^[A-Za-z0-9]+$/.test(strType),
    `Expected type name to match [A-Za-z0-9]+ but got '${strType}'${contextStr}`
  );
  return freeze({kind: 'NamedTypeReference', value: strType});
}

export default getType;
