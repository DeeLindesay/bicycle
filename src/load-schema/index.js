// @flow

import type {Schema} from '../flow-types';
import assert from 'assert';
import {readdirSync} from 'fs';
import freeze from '../utils/freeze';
import typeName from '../utils/type-name-from-value';
import BUILT_IN_SCALARS from '../scalars';
import suggestMatch from '../utils/suggest-match';
import normalizeObject from './normalize-object';
import normalizeScalar from './normalize-scalar';

const VALID_KEYS = ['scalars', 'objects'];
export default function loadSchema(input: Object): Schema {
  Object.keys(input).forEach(key => {
    if (VALID_KEYS.indexOf(key) === -1) {
      const suggestion = suggestMatch(VALID_KEYS, key);
      throw new Error(
        `Invalid key "${key}" in schema${suggestion}`
      );
    }
  });
  assert(
    Array.isArray(input.objects) && input.objects.every(obj => obj && typeof obj === 'object' && !Array.isArray(obj)),
    `Expected input.objects to be an Array<Object> but got ${typeName(input.objects)}`,
  );
  assert(
    input.scalars === undefined ||
    (
      Array.isArray(input.scalars) &&
      input.scalars.every(obj => obj && typeof obj === 'object' && !Array.isArray(obj))
    ),
    `Expected input.scalars to be an Array<Object> but got ${typeName(input.scalars)}`,
  );
  const types = {};
  const typeNames = [];
  let rootObject = null;
  input.objects.forEach(Type => {
    if (Type.name === 'Root') {
      rootObject = Type;
    } else if (Type.name) {
      if (typeNames.indexOf(Type.name) !== -1) {
        throw new Error(`Duplicate Object, "${Type.name}".  Each object & scalar must have a unique name.`);
      }
      typeNames.push(Type.name);
    }
  });
  if (input.scalars) {
    input.scalars.forEach(Scalar => {
      if (typeNames.indexOf(Scalar.name) !== -1) {
        throw new Error(`Duplicate Scalar, "${Scalar.name}".  Each object & scalar must have a unique name.`);
      } else if (Scalar.name) {
        typeNames.push(Scalar.name);
      }
    });
  }
  BUILT_IN_SCALARS.forEach(Scalar => {
    if (typeNames.indexOf(Scalar.name) === -1) {
      typeNames.push(Scalar.name);
    }
  });

  BUILT_IN_SCALARS.forEach(Scalar => {
    types[Scalar.name] = {
      kind: 'ScalarType',
      name: Scalar.name,
      description: Scalar.description,
      serialize: Scalar.serialize,
      parse: Scalar.parse,
    };
  });
  if (input.scalars !== undefined) {
    input.scalars.forEach(Scalar => {
      if (Scalar.name === 'Root') {
        throw new Error('You cannot have a scalar called Root');
      }
      types[Scalar.name] = normalizeScalar(Scalar);
    });
  }

  // objects have `id`s and a collection of `fields`
  input.objects.forEach(Type => {
    if (Type.name !== 'Root') {
      types[Type.name] = normalizeObject(Type, typeNames);
    }
  });
  if (!rootObject) {
    throw new Error('You must provide a Root object');
  }
  types.Root = normalizeObject(rootObject, typeNames);
  return freeze(types);
}

export function loadSchemaFromFiles(dirname: string): Schema {
  dirname = dirname.replace(/(\\|\/)$/, '');
  const schema = {objects: [], scalars: []};
  readdirSync(dirname + '/objects').forEach(filename => {
    // $FlowFixMe: intentionally not passing a literal to require
    let t = require(dirname + '/objects/' + filename);
    if (t.default) t = t.default;
    schema.objects.push(t);
  });
  let scalars = null;
  try {
    scalars = readdirSync(dirname + '/scalars');
  } catch (ex) {
    if (ex.code !== 'ENOENT') throw ex;
  }
  if (scalars) {
    scalars.forEach(filename => {
      // $FlowFixMe: intentionally not passing a literal to require
      let t = require(dirname + '/scalars/' + filename);
      if (t.default) t = t.default;
      schema.scalars.push(t);
    });
  }
  return loadSchema(schema);
}
