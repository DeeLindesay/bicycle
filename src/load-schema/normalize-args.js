// @flow

import type {ArgType} from '../flow-types';
import getType from './get-type';

function normalizeArgs(
  input: Object,
  typeName: string,
  fieldName: string,
  typeNames: Array<string>,
): {[key: string]: ArgType} {
  const result = {};
  Object.keys(input).forEach(argName => {
    result[argName] = {
      kind: 'arg',
      type: getType(input[argName], typeName + '.' + fieldName + '.' + argName, typeNames),
    };
  });
  return result;
}

export default normalizeArgs;
