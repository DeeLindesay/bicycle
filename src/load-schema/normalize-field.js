// @flow

import type {FieldType} from '../flow-types';
import assert from 'assert';
import freeze from '../utils/freeze';
import getTypeName from '../utils/type-name-from-value';
import suggestMatch from '../utils/suggest-match';
import getType from './get-type';
import normalizeArgs from './normalize-args';

const VALID_KEYS = ['type', 'description', 'args', 'resolve'];

function normalizeField(field: Object, fieldName: string, typeName: string, typeNames: Array<string>): FieldType {
  Object.keys(field).forEach(key => {
    if (VALID_KEYS.indexOf(key) === -1) {
      const suggestion = suggestMatch(VALID_KEYS, key);
      throw new Error(
        `Invalid key "${key}" in ${typeName}.${fieldName}${suggestion}`
      );
    }
  });
  assert(
    typeof field.type === 'string',
    `Expected ${typeName}.${fieldName}.type to be a string but got ${getTypeName(field.type)}`
  );
  assert(
    field.args === undefined || (field.args && typeof field.args === 'object'),
    `Expected ${typeName}.${fieldName}.args to be an object but got ${getTypeName(field.args)}`,
  );
  return freeze({
    type: getType(field.type, typeName + '.' + fieldName, typeNames),
    args: field.args ? normalizeArgs(field.args, typeName, fieldName, typeNames) : undefined,
    description: field.description,
    resolve: field.resolve,
  });
}

export default normalizeField;
