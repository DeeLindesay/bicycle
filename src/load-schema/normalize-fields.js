// @flow

import type {FieldType} from '../flow-types';
import assert from 'assert';
import freeze from '../utils/freeze';
import getTypeName from '../utils/type-name-from-value';
import normalizeField from './normalize-field';

function normalizeFields(fields: Object, typeName: string, typeNames: Array<string>): {[key: string]: FieldType} {
  const result = {};
  Object.keys(fields).forEach(fieldName => {
    const field = fields[fieldName];
    if (field === undefined) return;
    assert(
      /^[A-Za-z0-9]+$/.test(fieldName),
      `Expected ${typeName}'s field names to match [A-Za-z0-9]+ but got '${fieldName}'`,
    );
    assert(
      (field && typeof field === 'object') || typeof field === 'string',
      `Expected ${typeName}.${fieldName} to be an object or a string but got ${getTypeName(field)}`
    );
    result[fieldName] = normalizeField(
      typeof field === 'string' ? {type: field} : field,
      typeName,
      fieldName,
      typeNames,
    );
  });
  return freeze(result);
}

export default normalizeFields;
