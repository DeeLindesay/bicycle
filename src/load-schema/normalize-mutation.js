// @flow

import type {MutationType} from '../flow-types';
import assert from 'assert';
import suggestMatch from '../utils/suggest-match';
import getTypeName from '../utils/type-name-from-value';
import normalizeArgs from './normalize-args';
import getType from './get-type';

const VALID_KEYS = ['type', 'description', 'args', 'resolve'];

function normalizeMutation(
  mutation: Object,
  typeName: string,
  mutationName: string,
  typeNames: Array<string>,
): MutationType {
  Object.keys(mutation).forEach(key => {
    if (VALID_KEYS.indexOf(key) === -1) {
      const suggestion = suggestMatch(VALID_KEYS, key);
      throw new Error(
        `Invalid key "${key}" in ${typeName}.${mutationName}${suggestion}`
      );
    }
  });
  assert(
    mutation.description === undefined || typeof mutation.description === 'string',
    `Expected ${typeName}.${mutationName}.description to be a string but got ${getTypeName(mutation.description)}`,
  );
  assert(
    typeof mutation.resolve === 'function',
    `Expected ${typeName}.${mutationName} to have a resolve function`,
  );
  const type = mutation.type ? getType(mutation.type, typeName + '.' + mutationName, typeNames) : undefined;
  const args = mutation.args ? normalizeArgs(mutation.args, typeName, mutationName, typeNames) : undefined;
  return {
    type,
    description: mutation.description,
    args,
    resolve: mutation.resolve,
  };
}

export default normalizeMutation;
