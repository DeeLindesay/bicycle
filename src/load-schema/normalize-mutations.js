// @flow

import type {MutationsType} from '../flow-types';
import assert from 'assert';
import freeze from '../utils/freeze';
import getTypeName from '../utils/type-name-from-value';
import normalizeMutation from './normalize-mutation';

function normalizeMutations(mutations: Object, typeName: string, typeNames: Array<string>): MutationsType {
  const result = {
    set: mutations.set,
  };
  if (mutations.set !== undefined && typeof mutations.set !== 'function') {
    throw new Error(
      'The `set` mutation is a special case, it automatically takes `id`, `field` and `value`. ' +
      'You don\'t need to specify argument types for it so you should just use a function as shorthand. ' +
      `Look in ${typeName}.mutations.set to fix this.`
    );
  }
  Object.keys(mutations).forEach(mutationName => {
    if (mutationName === 'set') {
      return;
    }
    const mutation = mutations[mutationName];
    if (mutation === undefined) {
      return;
    }
    assert(
      /^[A-Za-z]+$/.test(mutationName),
      `Expected ${typeName}'s mutation names to match [A-Za-z0-9]+ but got '${mutationName}'`,
    );
    assert(
      typeof mutation === 'object',
      `Expected ${typeName}.${mutationName} to be an object but got ${getTypeName(mutation)}`
    );
    result[mutationName] = normalizeMutation(
      mutation,
      typeName,
      mutationName,
      typeNames,
    );
  });
  return freeze(result);
}

export default normalizeMutations;
