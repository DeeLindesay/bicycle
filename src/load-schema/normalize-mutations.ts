import assert from 'assert';
import normalizeMutation from './normalize-mutation';
import SchemaKind from '../types/SchemaKind';
import {Field, Mutation} from '../types/Schema';
import * as ta from './TypeAssertions';

function normalizeMutations(
  mutations: {},
  typeName: string,
  typeNames: string[],
  fields: {[name: string]: Field<any, any, any, any>},
): {[mutationName: string]: Mutation<any, any, any>} {
  // const result = {
  //   set: mutations.set,
  // };
  // if (mutations.set !== undefined && typeof mutations.set !== 'function') {
  //   throw new Error(
  //     'The `set` mutation is a special case, it automatically takes `id`, `field` and `value`. ' +
  //     'You don\'t need to specify argument types for it so you should just use a function as shorthand. ' +
  //     `Look in ${typeName}.mutations.set to fix this.`
  //   );
  // }
  const result: {[mutationName: string]: Mutation<any, any, any>} = {};
  const m = ta.Void
    .or(ta.AnyObject)
    .validate(mutations, typeName + '.mutations');
  if (m === undefined) {
    return {};
  }
  Object.keys(m).forEach(mutationName => {
    if (mutationName === 'set') {
      return;
    }
    const mutation = m[mutationName];
    if (mutation === undefined) {
      return;
    }
    assert(
      /^[A-Za-z]+$/.test(mutationName),
      `Expected ${typeName}'s mutation names to match [A-Za-z0-9]+ but got '${mutationName}'`,
    );
    if (mutationName === 'set') {
      result.set = {
        kind: SchemaKind.Mutation,
        name: 'set',
        description: 'Set a property on the given ' + typeName,
        argType: {
          kind: SchemaKind.Object,
          properties: {
            id: {
              kind: SchemaKind.Union,
              elements: [{kind: SchemaKind.String}, {kind: SchemaKind.Number}],
            },
            field: {
              kind: SchemaKind.Union,
              elements: Object.keys(fields).map(n => ({
                kind: SchemaKind.Literal,
                value: n,
              })),
            },
          },
        },
        resultType: {kind: SchemaKind.Void},
        auth: 'public',
        resolve(
          args: {id: number | string; field: string; value: any},
          ctx: any,
        ): void {},
      } as Mutation<any, any, any>;
    } else {
      result[mutationName] = normalizeMutation(
        mutation,
        typeName,
        mutationName,
        typeNames,
      );
    }
  });
  return result;
}

export default normalizeMutations;
