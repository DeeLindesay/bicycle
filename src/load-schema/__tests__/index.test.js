import {loadSchemaFromFiles} from 'bicycle/load-schema';

test('it can load a schema from a file', () => {
  expect(loadSchemaFromFiles(__dirname + '/../../test-schema')).toMatchSnapshot();
});
