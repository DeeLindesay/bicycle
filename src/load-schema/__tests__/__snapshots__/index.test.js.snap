exports[`test it can load a schema from a file 1`] = `
Object {
  "Root": Object {
    "description": undefined,
    "fields": Object {
      "todoById": Object {
        "args": Object {
          "id": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "string"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": Object {
          "kind": "NotNull",
          "type": Object {
            "kind": "NamedTypeReference",
            "value": "Todo"
          }
        }
      },
      "todos": Object {
        "resolve": [Function resolve],
        "type": Object {
          "kind": "NotNull",
          "type": Object {
            "kind": "List",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "Todo"
              }
            }
          }
        }
      }
    },
    "id": [Function idGetter],
    "kind": "NodeType",
    "mutations": Object {},
    "name": "Root"
  },
  "Todo": Object {
    "description": undefined,
    "fields": Object {
      "completed": Object {
        "type": Object {
          "kind": "NotNull",
          "type": Object {
            "kind": "NamedTypeReference",
            "value": "boolean"
          }
        }
      },
      "id": Object {
        "type": Object {
          "kind": "NotNull",
          "type": Object {
            "kind": "NamedTypeReference",
            "value": "id"
          }
        }
      },
      "title": Object {
        "type": Object {
          "kind": "NotNull",
          "type": Object {
            "kind": "NamedTypeReference",
            "value": "string"
          }
        }
      }
    },
    "id": [Function idGetter],
    "kind": "NodeType",
    "mutations": Object {
      "addTodo": Object {
        "args": Object {
          "completed": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "boolean"
              }
            }
          },
          "title": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "string"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": Object {
          "id": Object {
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "id"
              }
            }
          }
        }
      },
      "clearCompleted": Object {
        "args": Object {},
        "resolve": [Function resolve],
        "type": null
      },
      "destroy": Object {
        "args": Object {
          "id": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "id"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": null
      },
      "save": Object {
        "args": Object {
          "id": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "id"
              }
            }
          },
          "title": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "string"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": null
      },
      "toggle": Object {
        "args": Object {
          "checked": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "boolean"
              }
            }
          },
          "id": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "id"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": null
      },
      "toggleAll": Object {
        "args": Object {
          "checked": Object {
            "kind": "arg",
            "type": Object {
              "kind": "NotNull",
              "type": Object {
                "kind": "NamedTypeReference",
                "value": "boolean"
              }
            }
          }
        },
        "resolve": [Function resolve],
        "type": null
      }
    },
    "name": "Todo"
  },
  "boolean": Object {
    "description": "Either \`true\` or \`false\`",
    "kind": "ScalarType",
    "name": "boolean",
    "parse": [Function parse],
    "serialize": [Function serialize]
  },
  "id": Object {
    "description": undefined,
    "kind": "ScalarType",
    "name": "id",
    "parse": [Function parse],
    "serialize": [Function serialize]
  },
  "number": Object {
    "description": "Any floating point number",
    "kind": "ScalarType",
    "name": "number",
    "parse": [Function parse],
    "serialize": [Function serialize]
  },
  "string": Object {
    "description": "A string of arbitrary text",
    "kind": "ScalarType",
    "name": "string",
    "parse": [Function parse],
    "serialize": [Function serialize]
  }
}
`;
